### PR TITLE
Adding warm up for Zipformer2

### PR DESCRIPTION
**Pull request by manickavela29**
Hi @csukuangfj ,
### How and Where to use?

use '--warmp-up=<#warmup> for running warm ups for zipformer model'
    - Only supports Zipformer2 for now
    - Based on requirement it can be extended

### Why this is required?

1. It is a common practice to add warmup's for inference engines (especially with CUDA).
2. When the engine/server is live, first inference is very slow as it used for onnxruntime to create memory. Having warmup's will reduce the time.
3. Other execution providers of onnxruntime wlll also benefit from this.

### Data points :

> Without warmup : 

online-websocket-server.cc:main:79 Started!
online-websocket-server.cc:main:80 Listening on: 6007
online-websocket-server.cc:main:81 Number of work threads: 8
online-zipformer2-transducer-model.cc:RunEncoder:444 Encoder Duration : 0.457923ms
online-zipformer2-transducer-model.cc:RunEncoder:444 Encoder Duration : 0.170431ms
online-zipformer2-transducer-model.cc:RunEncoder:444 Encoder Duration : 0.167805ms
online-zipformer2-transducer-model.cc:RunEncoder:444 Encoder Duration : 0.168272ms
online-zipformer2-transducer-model.cc:RunEncoder:444 Encoder Duration : 0.173339ms
online-zipformer2-transducer-model.cc:RunEncoder:444 Encoder Duration : 0.169113ms
online-zipformer2-transducer-model.cc:RunEncoder:444 Encoder Duration : 0.167528ms
online-zipformer2-transducer-model.cc:RunEncoder:444 Encoder Duration : 0.169950ms
online-zipformer2-transducer-model.cc:RunEncoder:444 Encoder Duration : 0.175659ms

> With Warmup:

online-zipformer2-transducer-model.cc:RunEncoder:444 Encoder Duration : 0.524530ms
online-zipformer2-transducer-model.cc:RunEncoder:444 Encoder Duration : 0.190443ms
online-zipformer2-transducer-model.cc:RunEncoder:444 Encoder Duration : 0.188810ms
online-websocket-server-impl.cc:Run:250 Warm up completd : 3 times.
online-websocket-server.cc:main:79 Started!
online-websocket-server.cc:main:80 Listening on: 6007
online-websocket-server.cc:main:81 Number of work threads: 8
online-zipformer2-transducer-model.cc:RunEncoder:444 Encoder Duration : 0.193825ms
online-zipformer2-transducer-model.cc:RunEncoder:444 Encoder Duration : 0.169870ms
online-zipformer2-transducer-model.cc:RunEncoder:444 Encoder Duration : 0.169398ms
online-zipformer2-transducer-model.cc:RunEncoder:444 Encoder Duration : 0.172992ms
online-zipformer2-transducer-model.cc:RunEncoder:444 Encoder Duration : 0.172658ms
online-zipformer2-transducer-model.cc:RunEncoder:444 Encoder Duration : 0.173597ms
online-zipformer2-transducer-model.cc:RunEncoder:444 Encoder Duration : 0.173040ms
online-zipformer2-transducer-model.cc:RunEncoder:444 Encoder Duration : 0.171519ms
online-zipformer2-transducer-model.cc:RunEncoder:444 Encoder Duration : 0.172292ms

Let me know if additional change are required.
Thank you!

### CODE REVIEW

File: sherpa-onnx/csrc/online-recognizer-transducer-impl.h [hunk 1]
```diff
@@ -183,6 +184,39 @@ class OnlineRecognizerTransducerImpl : public OnlineRecognizerImpl {
            s->NumFramesReady();
   }
 
+  void WarmpUpRecognizer(int32_t warm_up,int32_t max_batch_size) {
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/online-recognizer-transducer-impl.h]:
```suggestion
  void WarmpUpRecognizer(int32_t warm_up,int32_t max_batch_size) const {
```

Also, please do an early return if `warm_up` is 0.

File: sherpa-onnx/csrc/online-recognizer.cc [hunk 2]
```diff
@@ -171,6 +171,10 @@ bool OnlineRecognizer::IsReady(OnlineStream *s) const {
   return impl_->IsReady(s);
 }
 
+void OnlineRecognizer::WarmpUpRecognizer(int32_t warm_up,int32_t max_batch_size) const {
+  impl_->WarmpUpRecognizer(warm_up,max_batch_size);
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/online-recognizer.cc]: Please invoke it only when `warm_up > 0`.

File: sherpa-onnx/csrc/online-model-config.h [hunk 3]
```diff
@@ -38,14 +39,15 @@ struct OnlineModelConfig {
                     const OnlineParaformerModelConfig &paraformer,
                     const OnlineWenetCtcModelConfig &wenet_ctc,
                     const OnlineZipformer2CtcModelConfig &zipformer2_ctc,
-                    const std::string &tokens, int32_t num_threads, bool debug,
+                    const std::string &tokens, int32_t num_threads, int32_t warm_up, bool debug,
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/online-model-config.h]: Please also change the python counterpart in sherpa-onnx/python/csrc/online-model-config.cc.

Comment (manickavela29) [reply sherpa-onnx/csrc/online-model-config.h]: yes,I have added the same in recent commit, 
please let me know if it is sufficient.

File: sherpa-onnx/csrc/online-recognizer.cc (hunk 2 above)
Comment (manickavela29) [reply sherpa-onnx/csrc/online-recognizer.cc]: yes, handled

File: sherpa-onnx/csrc/online-recognizer-transducer-impl.h (hunk 1 above)
Comment (manickavela29) [reply sherpa-onnx/csrc/online-recognizer-transducer-impl.h]: Handled !

**Comment by csukuangfj** [comment]
Could you fix the build errors reported by the CI?

**Comment by csukuangfj** [comment]
For instance, see the following error from
https://github.com/k2-fsa/sherpa-onnx/actions/runs/8674782410/job/23787499410#step:12:1178

```
In file included from /home/runner/work/sherpa-onnx/sherpa-onnx/sherpa-onnx/csrc/online-recognizer-impl.cc:9:
/home/runner/work/sherpa-onnx/sherpa-onnx/sherpa-onnx/csrc/online-recognizer-transducer-impl.h: In member function ‘void sherpa_onnx::OnlineRecognizerTransducerImpl::WarmpUpRecognizer(size_t, int32_t) const’:
/home/runner/work/sherpa-onnx/sherpa-onnx/sherpa-onnx/csrc/online-recognizer-transducer-impl.h:190:13: error: expected primary-expression before ‘chunk_size’
  190 |     int32_t chunk_size = model_->ChunkSize();
      |             ^~~~~~~~~~
/home/runner/work/sherpa-onnx/sherpa-onnx/sherpa-onnx/csrc/online-recognizer-transducer-impl.h:190:12: error: expected ‘;’ before ‘chunk_size’
  190 |     int32_t chunk_size = model_->ChunkSize();
      |            ^~~~~~~~~~~
      |            ;
/home/runner/work/sherpa-onnx/sherpa-onnx/sherpa-onnx/csrc/online-recognizer-transducer-impl.h:190:13: error: ‘chunk_size’ was not declared in this scope
  190 |     int32_t chunk_size = model_->ChunkSize();
      |             ^~~~~~~~~~
At global scope:
cc1plus: note: unrecognized command-line option ‘-Wno-missing-template-keyword’ may have been intended to silence earlier diagnostics
make[2]: *** [sherpa-onnx/csrc/CMakeFiles/sherpa-onnx-core.dir/build.make:776: sherpa-onnx/csrc/CMakeFiles/sherpa-onnx-core.dir/online-recognizer-impl.cc.o] Error 1
make[2]: *** Waiting for unfinished jobs....
```

File: sherpa-onnx/python/csrc/online-model-config.cc [hunk 4]
```diff
@@ -28,7 +28,7 @@ void PybindOnlineModelConfig(py::module *m) {
                     const OnlineParaformerModelConfig &,
                     const OnlineWenetCtcModelConfig &,
                     const OnlineZipformer2CtcModelConfig &, const std::string &,
-                    int32_t, bool, const std::string &, const std::string &>(),
+                    int32_t,int32_t, bool, const std::string &, const std::string &>(),
```
Comment (csukuangfj) [inline sherpa-onnx/python/csrc/online-model-config.cc]: Please also add
```
py::arg("warm_up") = 0
```
below.

Comment (manickavela29) [reply sherpa-onnx/python/csrc/online-model-config.cc]: Yes, added this.
updating a new commit with fixes for the build too.

**Comment by manickavela29** [comment]
review-requested:[@dspavankuar]

File: sherpa-onnx/csrc/online-recognizer-transducer-impl.h [hunk 5]
```diff
@@ -183,6 +184,43 @@ class OnlineRecognizerTransducerImpl : public OnlineRecognizerImpl {
            s->NumFramesReady();
   }
 
+  // Warmping up engine with wp: warm_up count and max-batch-size
+  void WarmpUpRecognizer(int32_t warmup, int32_t mbs) const {
+    auto max_batch_size = mbs;
+    if (warmup <= 0 || warmup > 100)
+      return;
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/online-recognizer-transducer-impl.h]:
```suggestion
    if (warmup <= 0 || warmup > 100) {
      return;
    }
```

File: sherpa-onnx/csrc/online-recognizer-transducer-impl.h [hunk 6]
```diff
@@ -183,6 +184,43 @@ class OnlineRecognizerTransducerImpl : public OnlineRecognizerImpl {
...
+    std::vector<std::vector<Ort::Value>> states_vec(max_batch_size);
+
+    auto memory_info =
+        Ort::MemoryInfo::CreateCpu(OrtDeviceAllocator, OrtMemTypeDefault);
+
+    std::array<int64_t, 3> x_shape{max_batch_size, chunk_size, feature_dim};
+
+    // filling with 0.0, so that it doesn't affact the states,
+    // and left_context which is maintained by the encoders
+    std::fill(features_vec.begin(), features_vec.end(), 0.0f);
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/online-recognizer-transducer-impl.h]:
```suggestion
```

`std::vector<float>` by default contains zeros on initialization, so there is no need to re-fill it with 0.

File: sherpa-onnx/csrc/online-recognizer.h [hunk 7]
```diff
@@ -162,6 +162,8 @@ class OnlineRecognizer {
     DecodeStreams(ss, 1);
   }
 
+  void WarmpUpRecognizer(int32_t warmup, int32_t mbs) const;
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/online-recognizer.h:172]: Please leave a comment describing what `mbs` is.

(I know it is short for max batch size, but I only know it after reading the code)

File: sherpa-onnx/csrc/online-websocket-server-impl.cc [hunk 8]
```diff
@@ -95,6 +95,11 @@ void OnlineWebsocketDecoder::InputFinished(std::shared_ptr<Connection> c) {
   c->eof = true;
 }
 
+void OnlineWebsocketDecoder::Warmup() {
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/online-websocket-server-impl.cc]:
```suggestion
void OnlineWebsocketDecoder::Warmup() const {
```

File: sherpa-onnx/csrc/online-websocket-server-impl.cc [hunk 9]
```diff
@@ -242,6 +247,23 @@ void OnlineWebsocketServer::Run(uint16_t port) {
   server_.set_reuse_addr(true);
   server_.listen(asio::ip::tcp::v4(), port);
   server_.start_accept();
+  auto recognizer_config = config_.decoder_config.recognizer_config;
+  int32_t warm_up = recognizer_config.model_config.warm_up;
+  std::string model_type = recognizer_config.model_config.model_type;
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/online-websocket-server-impl.cc]:
```suggestion
  const std::string &model_type = recognizer_config.model_config.model_type;
```

File: sherpa-onnx/csrc/online-websocket-server-impl.cc [hunk 10]
```diff
@@ -242,6 +247,23 @@ void OnlineWebsocketServer::Run(uint16_t port) {
...
+  if (0 < warm_up && warm_up < 100) {
+    if (model_type == std::string("zipformer2")) {
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/online-websocket-server-impl.cc]:
```suggestion
    if (model_type == "zipformer2") {
```

File: sherpa-onnx/csrc/online-websocket-server-impl.cc [hunk 11]
```diff
@@ -242,6 +247,23 @@ void OnlineWebsocketServer::Run(uint16_t port) {
...
+      decoder_.Warmup();
+      SHERPA_ONNX_LOGE("Warm up completd : %d times.", warm_up);
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/online-websocket-server-impl.cc]:
```suggestion
      SHERPA_ONNX_LOGE("Warm up completed : %d times.", warm_up);
```

File: sherpa-onnx/csrc/online-websocket-server-impl.cc [hunk 12]
```diff
@@ -242,6 +247,23 @@ void OnlineWebsocketServer::Run(uint16_t port) {
...
+    } else {
+      SHERPA_ONNX_LOGE("Only Zipformer2 has warmup support for now.");
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/online-websocket-server-impl.cc:258]:
```suggestion
      SHERPA_ONNX_LOGE("Only Zipformer2 has warmup support for now. Given: %s", model_type.c_str());
```

**Comment by manickavela29** [comment]
@pavankumar-ds

File: sherpa-onnx/csrc/online-websocket-server-impl.h [hunk 13]
```diff
@@ -85,6 +85,8 @@ class OnlineWebsocketDecoder {
   // signal that there will be no more audio samples for a stream
   void InputFinished(std::shared_ptr<Connection> c);
 
+  void Warmup();
```
Comment (csukuangfj) [inline sherpa-onnx/csrc/online-websocket-server-impl.h]:
```suggestion
  void Warmup() const;
```

File: sherpa-onnx/python/csrc/online-model-config.cc [hunk 14]
```diff
@@ -27,14 +27,16 @@ void PybindOnlineModelConfig(py::module *m) {
...
+                    const OnlineZipformer2CtcModelConfig &,
+                    const std::string &, int32_t, int32_t,
+                    bool, const std::string &, const std::string &>(),
            py::arg("transducer") = OnlineTransducerModelConfig(),
            py::arg("paraformer") = OnlineParaformerModelConfig(),
            py::arg("wenet_ctc") = OnlineWenetCtcModelConfig(),
            py::arg("zipformer2_ctc") = OnlineZipformer2CtcModelConfig(),
-           py::arg("tokens"), py::arg("num_threads"), py::arg("debug") = false,
-           py::arg("provider") = "cpu", py::arg("model_type") = "")
+           py::arg("tokens"), py::arg("warm_up") = 0, py::arg("num_threads"),
```
Comment (csukuangfj) [inline sherpa-onnx/python/csrc/online-model-config.cc]: please follow the order at
https://github.com/k2-fsa/sherpa-onnx/blob/bdb74815d8a0b9dd427a0614453050b78b3243f6/sherpa-onnx/csrc/online-model-config.h#L21-L24

It should be
```cpp
py::arg("tokens"),
py::arg("num_threads") = 1,
py::arg("warm_up") = 0,
py::arg("debug") = false,
```

**Review by csukuangfj — COMMENTED** [review]
Thanks!

Left some minor comments.

File: sherpa-onnx/csrc/online-recognizer-transducer-impl.h (hunk 5 above)
Comment (csukuangfj) [reply sherpa-onnx/csrc/online-recognizer-transducer-impl.h]: We use such a code style that it also adds `{}` to a `if` statement even if it contains a single `return`.

Comment (manickavela29) [reply sherpa-onnx/csrc/online-recognizer-transducer-impl.h]: ok thanks

File: sherpa-onnx/csrc/online-recognizer-transducer-impl.h (hunk 6 above)
Comment (manickavela29) [reply sherpa-onnx/csrc/online-recognizer-transducer-impl.h]: ok got it

File: sherpa-onnx/csrc/online-recognizer.h (hunk 7 above)
Comment (manickavela29) [reply sherpa-onnx/csrc/online-recognizer.h:172]: yes, will add this

File: sherpa-onnx/python/csrc/online-model-config.cc (hunk 14 above)
Comment (manickavela29) [reply sherpa-onnx/python/csrc/online-model-config.cc]: Taken care.

**Review by csukuangfj — APPROVED** [review]
Thank you for your contribution!

**Comment by manickavela-uni** [comment]
Thank you  😄

**Comment by manickavela29** [comment]
Thank you 😄

**Pull request merged**